cmd/kubeadm/app/phases/addons/addons_test.go

```diff
@@ -86,3 +92,69 @@ func TestCompileManifests(t *testing.T) {
 		}
 	}
 }
+
+// This unit test exists because tolerations weren't decoded correctly by kuberuntime.DecodeInto
+// In case that regression ever happens again, this unit test will catch it
+func TestTolerationsPreserved(t *testing.T) {
+	var tests = []struct {
+		manifest string
+		data     interface{}
+		f        func(data []byte) []v1.Toleration
+		expected []v1.Toleration
+	}{
+		{
+			manifest: KubeProxyDaemonSet,
+			data: struct{ Image, ClusterCIDR, MasterTaintKey string }{
+				Image:          "foo",
+				ClusterCIDR:    "- foo",
+				MasterTaintKey: kubeadmconstants.LabelNodeRoleMaster,
+			},
+			f: func(data []byte) []v1.Toleration {
+				kubeproxyDaemonSet := &extensions.DaemonSet{}
+				if err := kuberuntime.DecodeInto(api.Codecs.UniversalDecoder(), data, kubeproxyDaemonSet); err != nil {
```

Comment:
I don't understand how this can work. k8s.io/kuberentes/pkg/api.Codecs shouldn't understand k8s.io/client-go types (e.g., kubeproxyDaemonSet). Maybe the UniversalDecoder() doesn't bother to infer the type of kubeproxyDaemonSet?

Apart from that, why does kubeadm need to use api.Codecs? Does kubeadm convert object to the internal versions (e.g.,pkg/api#Pod)?